web/client/plugins/Widgets.jsx

```diff
@@ -217,8 +219,23 @@ compose(
             ),
             withPropsOnChange(
                 ['activeWidget', 'isSingleWidgetLayout', 'widgets'],
-                ({activeWidget, dropdownWidgets, setActiveWidget, isSingleWidgetLayout, widgets, toolsOptions, layouts}) => {
+                ({
+                    activeWidget,
+                    dropdownWidgets,
+                    isSingleWidgetLayout,
+                    widgets,
+                    toolsOptions,
+                    layouts,
+                    setActiveWidget,
+                    toggleCollapseAll: collapseAll
+                }) => {
                     if (activeWidget && isSingleWidgetLayout && widgets.length) {
+                        const showWidget = widgets.find(el => el.id === activeWidget.id);
+                        if (!showWidget) {
+                            // If single widget should not be displayed - hide all other widgets to make widgets tray
+                            // update its state properly (show "Expand all widgets" button)
+                            collapseAll();
+                        }
```

Comment:
I think we cannot add a side effect related to another plugin inside the render lifecycle of a component, this needs to be changed. Could you explain better what we are trying to fix here?

Reply:
Could we provide a new toggleCollapse function from the Widgets plugin to the [WidgetsView](https://github.com/geosolutions-it/MapStore2/blob/master/web/client/components/widgets/view/WidgetsView.jsx#L119) instead to check if the active widget is available in the widgets array and trigger a side effect?

I did not test this, it just to give an idea on what I meant (I'm not sure it's possible to use the same name toggleCollapse for the handler )
```js
withHandlers({
    toggleCollapse: props => (w) => {
        const showWidget = props.widgets?.find(el => el.id === props.activeWidget?.id);
        if (props.isSingleWidgetLayout && showWidget) {
            props.collapseAll();
        } else {
            props.toggleCollapse(w);
        }
    }
})
```

or if this does not cover all the requirements we could use lifecycle enhancer instead with componentDidUpdate or componentWillUpdate